File: go/src/prime_factors/prime_factors.go
package prime_factors

func Factors(num int) []int {
  return FactorsRec(num, 2)
}

func FactorsRec(num int, candidate int) []int {
  if num < candidate  {
    return []int{}
  }

  if num % candidate == 0 {
    return append([]int{candidate}, Factors(num / candidate)...)
  } else {
    return FactorsRec(num, candidate + 1)
  }
}

File: scala/src/main/scala/PrimeFactors.scala
package com.crebma.primeFactors

object PrimeFactors {

  def factors(num: Int, candidate: Int = 2) : List[Int] = {
    if(num < candidate) {
      List[Int]()
    } else if (num % candidate == 0) {
      List[Int](candidate) ++ factors(num / candidate)
    } else {
      factors(num, candidate + 1)
    }
  }

}
